doc/man3/EZ_CreateItem.3
'\"
'\" Copyright (c) 1997 Maorong Zou
'\" 
.TH EZ_CreateItem 3 "" EZWGL "EZWGL Functions"
.BS
.SH NAME
EZ_CreateItem \- create a display item
.SH SYNOPSIS
.nf
.B #include <EZ.h>
.sp
.BI "EZ_Item *EZ_CreateItem(int " type ", " ... )

.SH ARGUMENTS
\fItype\fR Specifies a symbolic item type. It must be one
of EZ_LABEL_ITEM or EZ_WIDGET_ITEM.
.sp
\fI ...\fR Specifies configuration options, a list of symbols 
each followed by one or more values. The last symbol must
be \fBNULL\fR, which is itself not followed by any values.
.sp
.SH DESCRIPTION
.PP
\fIEZ_CreateItem\fR creates a display item and initializes
the item using the specified configuration options. For example
.sp
.nf
   extern EZ_CallBack  openXterm(EZ_Item *, void *);
   EZ_Item *item;
                        
   item = EZ_CreateItem(EZ_LABEL_ITEM,      /* type */
                        /* option,       option value(s) */
                        EZ_LABEL_STRING, "xterm",
                        EZ_ORIENTATION,   EZ_VERTICAL,
                        EZ_PIXMAP_FILE,  "xterm.xpm",
                        EZ_CALLBACK,      openXterm, NULL,
                        0);
.sf
.sp
creates a label item, labled by the image "xterm.xpm" on top of
the string 'xterm'. It also associates a callback 'openXterm'
procedure to the item.
.PP
Currently, only two types of items are supported.
.PP
.SH CONFIGURATION SYMBOLS AND THEIR VALUES
The set of item configuration options and their values
are listed below. For each option value, the string before
the last _ is its type.
.sp
\fBEZ_BORDER_WIDTH\fR, \fIint_bw\fR 
.in +5
Specifies a border width.
.in -5
.sp
\fBEZ_BORDER_TYPE\fR, \fIint_bt\fR
.in +5
Specifies a border type. \fIint_bt\fR must be one of EZ_BORDER_NONE,
EZ_BORDER_FLAT, EZ_BORDER_RIDGE, EZ_BORDER_GROOVE, EZ_BORDER_RAISED or
EZ_BORDER_SUNKEN.
.in -5
.sp
\fBEZ_PADX\fR, \fIint_px\fR
.in +5
Specifies the horizontal padding between the item border and its contents.
.in -5
.sp
\fBEZ_PADY\fR, \fIint_py\fR
.in +5
Specifies the vertical padding between the item border and its contents.
.in -5
.sp
\fBEZ_IPADX\fR, \fIint_ipx\fR
.in +5
Specifies the amount of spaces to be inserted horizontally between
textual label and image label. It is used only when the item has
a horizontal orientation.
.in -5
.sp
\fBEZ_IPADY\fR, \fIint_ipy\fR
.in +5
Specifies the amount of spaces to be inserted vertically between
textual label and image label. It is used only when the item has
a vertical orientation.
.in -5
.sp
\fBEZ_ORIENTATION\fR, \fIint_ori\fR
.in +5
Specifies the orientation of an item. \fIori\fR must be EZ_HORIZONTAL
or EZ_VERTICAL.
.in -5
.sp
\fBEZ_LABEL_POSITION\fR, \fIint_lp\fR
.in +5
Specifies a location to anchor a label. \fIint_lp\fR must be one of
EZ_CENTER, EZ_LEFT, EZ_RIGHT, EZ_TOP, EZ_BOTTOM,  EZ_TOP_LEFT,
EZ_TOP_RIGHT, EZ_BOTTOM_LEFT or EZ_BOTTOM_RIGHT.  This attribute
is used only when an item has been given a size which is larger
than its minimal size.
.in -5
.sp
\fBEZ_LABEL_JUSTIFICATION\fR, \fIint_jst\fR
.in +5
Specifies the justfication for textual labels. \fIint_jst\fR must be
one of EZ_CENTER, EZ_LEFT or EZ_RIGHT.
.in -5
.sp
\fBEZ_CALLBACK\fR, \fIEZ_CallBack_cbk\fR, \fIvoidptr_cdata\fR
.in +5
Specifies a callback procedure and a client data to be passed to the
callback. Use this option only when an item is to have only one callback.
.in -5
.sp
\fBEZ_FOREGROUND\fR, \fIstr_fg\fR
.in +5
Specifies a forground color.
.in -5
.sp
\fBEZ_BACKGROUND\fR, \fIstr_bg\fR
.in +5
Specifies a background color.
.in -5
.sp
\fBEZ_PIXMAP_FILE\fR, \fIstr_filename\fR
.in +5
Specifies an X bitmap file or a X pixmap file (an xpm image).
.in -5
.sp
\fBEZ_BITMAP_DATA\fR, \fIarray_of_char\fR,\fIint_width\fR,\fIint_height\fR
.in +5
Specify an X bitmap.
.in -5
.sp
\fBEZ_PIXMAP_DATA\fR, \fIarray_of_str\fR
.in +5
Specifies an XPM image data.
.in -5
.sp
\fBEZ_X_PIXMAP\fR,
\fIPixmap_p\fR,\fIint_x\fR,\fIint_y\fR,\fIint_width\fR,\fIint_height\fR
.in +5
Specify a rectangular region of a native X pixmap to be used as an
image label.
.in -5
.sp
\fBEZ_LABEL_PIXMAP\fR, \fIEZ_Pixmap_p\fR
.in +5
Specifies an internal pixmap.
.in -5
.sp
\fBEZ_FONT_ID\fR, \fIint_id\fR
.in +5
Specifies a fond id.
.in -5
.sp
\fBEZ_FONT_NAME\fR, \fIstr_filename\fR
.in +5
Spcifies an X11 font name.
.in -5
.sp
\fBEZ_TEXT_LINE_LENGTH\fR, \fIint_len\fR
.in +5
Specifies the wrap around length for textual labels.
.in -5
.sp
\fBEZ_IMAGE_FILE\fR, \fIstr_name\fR
.in +5
Specifies an image file name. Supported image formats are: ppm, xpm,
bmp, gif and jpeg. 
.in -5
.sp
\fBEZ_LABEL_STRING\fR, \fIstr_label\fR
.in +5
Specifies a textual label.
.in -5
.sp
\fBEZ_CLIENT_PTR_DATA\fR, \fIptr_data\fR
.in +5
Specifies a generic pointer to be stored in a item.
.in -5
.sp
\fBEZ_CLIENT_INT_DATA\fR, \fIint_data\fR
.in +5
Specifies an integer to be stored in a item.
.in -5
.sp
\fBEZ_DESTROY_CALLBACK\fR, \fIEZ_CallBack_callback\fR, \fIptr_data\fR
.in +5
Specify a destroy callback and a client data to be passed to the
callback. Use this option only when an item is to have only one
destroy callback.
.in -5
.sp
\fBEZ_DND_DRAG_CURSOR\fR, \fICursor_cur\fR
.in +5
Specifies an X cursor for drag and drop. The specified cursor will
replace the default cursor when the item is being draged.
.in -5
.sp
\fBEZ_DND_BUBBLE_HELP\fR, \fIstr_hlp\fR
.in +5
Specifies a DND bubble help string. DND bubble will be displayed when
a drag is paused over an item and the item is a drop site. 
.in -5
.sp
\fBEZ_X\fR, \fIint_x\fR 
.br
.in +5
Specifies the x coordinate of the upper-left
corner of an item in its hosting window.
.in -5
.sp 
\fBEZ_Y\fR, \fIint_y\fR 
.in +5
Specifies the y coordinate of the upper-left
corner of an item in its hosting window.
.in -5
.sp
\fBEZ_WIDTH\fR, \fIint_w\fR
.in +5
Specifies the width of an item.
.in -5
.sp
\fBEZ_HEIGHT\fR, \fIint_h\fR
.in +5
Specifies the height of an item.
.in -5
.sp
.sp
\fBEZ_LOCATION\fR, \fIint_x\fR, \fIint_y\fR
.in +5
Specify the coordinates of the upper-left corner of an item
in its hosting window.
.in -5
.sp
\fBEZ_SIZE\fR, \fIint_w\fR, \fIint_h\fR
.in +5
Specify the width and height of an item.
.in -5
.sp
.SH "SEE ALSO"
EZ_ConfigureItem(3), EZ_DestroyItem(3)




ChangeLog
Fri Dec 10 23:44:16 1999  m z  <zou@hitachi1.ma.utexas.edu>
        
        * Added EZ_SSheetSortRegionR, EZ_SSheetSortColumns/Rows
        
1999-12-04  Maorong Zou  <mzou@linux115.ma.utexas.edu>

        * Released version 1.50 
        * added a CVS repository entry 
        *  :pserver:anonymous@cvs.math.utexas.edu:/home/cvs/EZWGL
        * and a read-only cvs web interface
        *  http://cvs.math.utexas.edu/cgi-bin/cvswe.cgi?cvsroot=EZWGL

1999-11-18  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	* Added THUMB_WHEEN and SPREAD_SHEET widget
	* a few modifications in in FigItem ...

1999-09-22  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	* EZ_RESOURCE_STRING resource
	* EZ_LedLoadPixelMatrix() EZ_LedLoadOnOffMatrix()
	* EZ_AllowDualLabel()
	* EZ_FlashWidgetBorder2()
	* NewFile  EZ_ResourceEdit.c
	* Incompatible change:  EZ_LED_PIXEL_SIZE now
	*        takes a pair of ints (xsize, ysize)

	
Sun Aug 22 01:26:59 1999  m z  <mzou@crab.ma.utexas.edu>

        * Separator Widget

Tue Aug 17 02:10:03 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added   EZ_SlideOpenWidget,  EZ_SlideCloseWidget
        *         EZ_SlideOpenWidgetRelativeTo
        * new widget EZ_WIDGET_HISTOGRAM

Tue Aug 10 12:33:26 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added the pare EZ_FreezeDisplay() and EZ_ThawDisplay()
        * New File EZ_ResourceEdit.c (resource editor functions)
        * Added EZ_RegisterApplicationResources(n,spec,config)
        * Modified all resource handling functions
        * changed EZ_Initialize and EZ_InitializeXrm from void to int
        *   (return -1 on failure, 0 on success)
        
Mon Jul 12 02:13:37 1999  m z  <mzou@crab.ma.utexas.edu>

        * New widget GradientBar
        * Fix a bug in EZ_PixelValue2RGB
        
Tue Jul  6 02:17:29 1999  m z  <mzou@crab.ma.utexas.edu>

        * Fix buffer overflow bug in EZ_Xpm.c
        * added line/word wrap for textWidget, IN BROWSING FILE MODE ONLY!!
        * (need to rewrite textWidget, the current line based
        *  implementation is a design flaw)

Thu Jun 17 01:49:17 1999  m z  <mzou@crab.ma.utexas.edu>

        * New resources: SHAPE_PIXMAP, SHAPE_FILE
        * EZ_Bitmap *EZ_CreateShapeMaskFromFigure(fitem) EZ_Item *fitem;
        * Pixmap EZ_MakePixmapFromFigItem(item, width, height)
             EZ_Item *item; int *width, *height;
        
Fri May 14 23:30:19 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added a new tree type: branched-list-tree

Thu May 13 01:50:55 1999  m z  <mzou@crab.ma.utexas.edu>

        * New Widget:  ScrollButton, SpinButton, Locator

Sun May  9 02:03:17 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added UDF widget resource handling.
        * EZ_RegisterResourcesAll(type, nresrc, resrc, configure);
        * EZ_RegisterResourcesWidget(widget, nresrc, resrc, configure);        

Tue May  4 23:36:31 1999  m z  <mzou@crab.ma.utexas.edu>

        *  Widget (horizontal/vertical) ruler
        *  
        *  Configuration options:
        *  EZ_HIGHLIGHT_BACKGROUND, EZ_HIGHLIGHT_FOREGROUND,
        *  EZ_OVERRIDE_REDIRECT

Tue May  4 01:22:57 1999  m z  <mzou@crab.ma.utexas.edu>

        * New Function 
        * void  EZ_ConstrainFreeLabel(EZ_Widget *widget, int flag);
        * void  EZ_MoveWidgetWindow(EZ_Widget *widget, int x, int y);
        * void  EZ_ResizeWidgetWindow(EZ_Widget *widget, int w, int h);
        * void  EZ_MoveResizeWidgetWindow(EZ_Widget *widget, int x, int y, int w, int h);        
        * Added optimization for manipulating images.
        *
        
Sat Apr 17 01:25:25 1999  m z  <mzou@crab.ma.utexas.edu>

        * Autoconf (finally)

1999-04-16  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * added a vt100 emulator widget (EZ_WIDGET_TERM), derived from xiterm.
        * added support for image format: png and tiff
        
Tue Dec  8 11:39:04 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        *  new function:     EZ_ScaleRGBT
        *                    EZ_XPixmap2RGB
        *                    EZ_SelectInput
        *                    EZ_CreateBackgroundPixmapsFromRawRGB

Sat Dec  5 03:06:58 1998  m z  <mzou@bob.ma.utexas.edu>

        * EZ_LABEL_SHADOW  "labelShadow" config. option

Thu Dec  3 14:54:21 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * new function:   EZ_SetDirTreeSorter
        *                 EZ_MarkWidgetDirty
        
Mon Nov 30 12:25:50 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * EZ_WM_IGNORE  "wmIgnore"    configuration option
        *
        * new function    EZ_GetIconTextGeometry
        *                 EZ_SetGlobalEventFilter(filter, data)
        
Thu Nov 19 16:35:47 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	*  EZ_TILE_ORIGIN  tileOrigin: mode x y
	*  resources.  mode=0,1,2    (0 for off)

	*  It is now possible to make all EZ widgets
	*  pseudo transparent under Enlightment WM ...

	*  A few image readers have been modified to
	*  return RGB images.

	*  new functions EZ_ScaleRGB, EZ_ReadImageFile,
	*                EZ_ScaleLabelPixmap,
	*                EZ_SwapLabelPixmaps, 
	*                EZ_CopyLabelPixmaps,
	*                EZ_TraverseAllWidgets,

	*  In EZ_DisplayWidget, the order for displaying widget tree
	*  is reversed. Now widgets refresh a lot faster visually.
	
	*  Bug Fix:  whe a radio button is destroyed, the widget is
	*            not removed from its button group immediately.



lib/EZ_CreateItem.c
/*********************************************************************
 *
 *         EZWGL, the EZ Widget and Graphics Library
 *
 * *             Copyright (C) 1996, 1997, 1998  Maorong Zou
 *  
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 **********************************************************************/
/*
 *  June 1996.  Beta Release.
 *  Sept 1996.  Release Version 1.0
 *  Dec 1996.  Release Version 1.1 Beta
 *  April 1997.  Release Version 1.2
 *  November 1997.  Release Version 1.3
 *  November 1998.  Release Version 1.4
 *  December 1999.  Release Version 1.50
 */
/*****************************************************************
 *
 *  Display items
 *
 *****************************************************************/
#define _EZ_CREATE_ITEM_C_

#include "EZ_Widget.h"
#include "EZ_WidgetConfig.h"


/*****************************************************************
 *
 * public functions
 */
EZ_Item *EZ_CreateItemXrm MY_VARARGS(int,  the_args)
     MY_VARARGS_DECL
{
  va_list         ap;
  EZ_Item         *item = NULL;
  int              type;
  char      *iname,  *cname;
  XrmQuark  qiname,  qcname; 

  type = (int)( MY_VARARGS_START(int, the_args, ap));
  iname = (char *)va_arg(ap, char *);
  cname = (char *)va_arg(ap, char *);
  if(iname != NULL) qiname = EZ_StringToQuark(iname);
  else qiname = NULLQUARK;
  if(cname != NULL) qcname = EZ_StringToQuark(cname);
  else qcname = NULLQUARK;

  switch(type)
    {
    case EZ_LABEL_ITEM:
      item = EZ_CreateLabelItem(NULL, NULL);
      break;
    case EZ_WIDGET_ITEM:
      item = EZ_CreateWidgetItem(NULL);
      break;
    case EZ_FIG_ITEM:
      item = EZ_CreateFigItem();      
      break;
    default:
      item = EZ_CreateUDFItem(type);
      break;
    }

  if(item)
    {
      if(qiname != NULLQUARK) EZ_ItemIName(item) = qiname;
      if(qcname != NULLQUARK) EZ_ItemCName(item) = qcname;
      
#include "EZ_ItemAttr.h"

      /* resources */
      EZ_GetAndUseItemResources(item);
    }
  return(item);
}
/*****************************************************************
 *
 * public functions
 */
EZ_Item *EZ_CreateItem MY_VARARGS(int,  the_args)
     MY_VARARGS_DECL
{
  va_list         ap;
  EZ_Item         *item = NULL;
  int              type;

  type = (int)( MY_VARARGS_START(int, the_args, ap));

  switch(type)
    {
    case EZ_LABEL_ITEM:
      item = EZ_CreateLabelItem(NULL, NULL);
      break;
    case EZ_WIDGET_ITEM:
      item = EZ_CreateWidgetItem(NULL);
      break;
    case EZ_FIG_ITEM:
      item = EZ_CreateFigItem();      
      break;
    default:
      item = EZ_CreateUDFItem(type);
      break;
    }

  if(item)
    {
#include "EZ_ItemAttr.h"
      /* resources */
      EZ_GetAndUseItemResources(item);
    }
  return(item);
}
/*****************************************************************/
#undef _EZ_CREATE_ITEM_C_
